Copy documentation from temp repo

baijum opened the pull request:
Signed-off-by: Baiju Muthukadan <baiju.m.mail@gmail.com>

Copied from: https://github.com/slemeur/service-binding-documentation

Srivaralakshmi reviewed (changes requested):
@baijum PTAL at the following suggestions and kindly address them all:

- content related to `.spec.mountPath` needs to be taken out.
- instances of 'injected', 'injection', 'injecting', and 'injects' needs to be changes to 'projected', 'projection', 'projecting', and 'projects'. 
- Check for all similar instances of injection terminology and change to projection terminology.
- bindings ---> binding metadata (check all instances and change globally)

Thanks!

baijum commented:
> @baijum PTAL at the following suggestions and kindly address them all:
> 
> * content related to `.spec.mountPath` needs to be taken out.
> * instances of 'injected', 'injection', 'injecting', and 'injects' needs to be changes to 'projected', 'projection', 'projecting', and 'projects'.
> * Check for all similar instances of injection terminology and change to projection terminology.
> * bindings ---> binding metadata (check all instances and change globally)
> 
> Thanks!

I have addressed these suggestions.

kadel commented on docs/docs/creating-service-bindings/creating-service-binding.md:
```diff
@@ -0,0 +1,71 @@
+You also required the application ready before creating the ServiceBinding
+resource.
+
+There are two APIs supported by the Service Binding Operator.  Both has
```
It would be nice if documentation could expand on what are the main difference between those APIs and which one is recommended.

baijum replied:
Added a comparison. 481088d985e0b477230270d6234efa8d855e43c5

baijum commented:
> should /docs/docusaurus.config.js be removed too?

This is required for docusaurus tool to generate the website. If we decide to use asciidoc, we can remove it.

Srivaralakshmi commented:
@baijum @pmacik 

This prerequisite is missing under the getting-started/quick-start.md file. Please add it:

- You have installed PostgreSQL `psql` CLI.

Thanks!

baijum commented:
> @baijum @pmacik
> 
> While testing step 3: Initialize the database with the schema and sample data, for the Creating a PostgreSQL database instance procedure, I am getting the output: "bash: line 12: psql: command not found".
> 
> This prerequisite is missing under the getting-started/quick-start.md file. Line 44 onwards. Please add it:
> 
> * You have installed PostgreSQL `psql` CLI.
> 
> Thank you so much for confirming my doubts and helping me complete the testing successfully!

Done.

pedjak commented:
why is `docs/pull_request_template.md` removed in this PR?

baijum commented:
> why this PR removes:
> 
> * docs/ci.md
> * docs/community-release.md

Restored.

> Also, the directory where the userguide is located is `docs/docs` - would not be better if it is `docs/userguide` so that user can easily navigate to it?

We are using `docusaurus` tool. So, moved all under the userguide directory.

baijum commented:
> why is `docs/pull_request_template.md` removed in this PR?

Restored.

pedjak reviewed (changes requested):
general remarks
* spell check should be run on the doct
* replace in general "binding metadata" with "binding data"
* replace "service binding custom resource" with "service binding resource"
* spec API meanwhile got updated to `servicebinding.io` and version is `v1alpha3`

jasperchui commented:
A lot of the commands reference https://github.com/slemeur/service-binding-documentation this should be updated to reflect the new location.

pmacik commented on docs/userguide/docs/getting-started/quick-start.md:
````diff
@@ -0,0 +1,342 @@
+running the following command:
+
+```bash
+curl -sL https://github.com/slemeur/service-binding-documentation/raw/main/static/resources/setup-pg-sbo-rbac.sh | bash
````
or something like:

```suggestion
kubectl apply -f - << EOD
apiVersion: rbac.authorization.k8s.io/v1
kind: ClusterRole
metadata:
  name: postgrescluster-reader
  labels:
    servicebinding.io/controller: "true"
rules:
- apiGroups: ["postgres-operator.crunchydata.com"]
  resources: ["postgresclusters"]
  verbs: ["get", "watch", "list"]
EOD
```

baijum commented on docs/docs/samples/sample-1.md:
```diff
@@ -0,0 +1,5 @@
+---
```
Removed.

baijum commented on docs/userguide/docs/intro.md:
```diff
@@ -0,0 +1,71 @@
+  <dt>Service Binding</dt>
+  <dd>The act of or representation of the action of providing information about a Service to an Application</dd>
+
+</dl>
```
Changed to markdown

baijum commented on docs/userguide/docs/getting-started/quick-start.md:
```diff
@@ -0,0 +1,342 @@
+Service Binding manages the data plane for applications and backing services.
+This guide provides the minimal information a needs to install the Service
+Binding Operator and use it to create a binding connection between components
+within an application.
```
Removed the paragraph.

baijum commented on docs/userguide/docs/exposing-binding-data/intro-expose-binding.md:
```diff
@@ -0,0 +1,86 @@
+   `sliceOfMaps`.  Defaults to `string` if `elementType` is not specified.
+
+* `objectType`: Specifies whether the value of the element indicated in `path`
+  refers to a `ConfigMap`, `Secret` or a plain string in the current namespace!
```
removed exclamation mark.

baijum commented on docs/userguide/docs/exposing-binding-data/olm-descriptors.md:
```diff
@@ -0,0 +1,245 @@
+Bundle](https://github.com/operator-framework/operator-registry/blob/master/docs/design/operator-bundle.md).
+You can add [OLM
+descriptors](https://github.com/openshift/console/blob/master/frontend/packages/operator-lifecycle-manager/src/components/descriptors/reference/reference.md)
+to describe the binding metadata that is to be exposed.  The OLM
```
Updated the sentence with reason.

baijum commented on docs/userguide/docs/exposing-binding-data/olm-descriptors.md:
```diff
@@ -0,0 +1,245 @@
+
+The Service Binding Operator detects the OLM descriptors.  Then the Service
+Binding will create a Secret with the values extracted based on the same.
+Finally the Service Binding will project the values into the application.
```
removed implementation details.

baijum commented on docs/userguide/docs/exposing-binding-data/olm-descriptors.md:
````diff
@@ -0,0 +1,245 @@
+
+```
+urn:alm:descriptor:io.kubernetes:ConfigMap
+```
````
Done.

baijum commented on docs/userguide/docs/exposing-binding-data/adding-annotation.md:
```diff
@@ -0,0 +1,316 @@
+values into the application.
+
+The following examples show the annotations that are added under the metadata
+section and a referenced ConfigMap object from a resource:
```
Updated sentence to include Secret also

baijum commented on docs/userguide/docs/getting-started/quick-start.md:
```diff
@@ -0,0 +1,342 @@
+## Create a PostgreSQL database instance
+
+The application is going to use to a PostgreSQL database backend which can be
+setup using the `v5` of [Crunchy PostgreSQL operator from OperatorHub.io][crunchy].
```
v5 is the channel name. Updated text accordingly.

baijum commented on docs/userguide/docs/getting-started/quick-start.md:
```diff
@@ -0,0 +1,342 @@
+We just need to make sure that the operator is availabe in our `my-postgresql` namespace.
+
+Additionally, we need to allow the Service Binding Operator to `read`, `watch` and `list` Crunchy PostgreSQL operator's
+custom resources by creating appropriate `ClusterRole` and `ClusterRoleBinding` resources. You can do it by
```
Done

baijum commented on docs/userguide/docs/getting-started/quick-start.md:
````diff
@@ -0,0 +1,342 @@
+service.binding/type: path={.metadata.annotations.type}
+```
+
+The annotions points to the values of `database`, `host`, `port` and `username` in resource
````
Done

baijum commented on docs/userguide/docs/getting-started/quick-start.md:
````diff
@@ -0,0 +1,342 @@
+schema and we will also project a sample data set, so we can play around with the
+application.
+
+3. You can initialize the database with the schema and sample data using this
+   command:
+
+```bash
+curl -sL https://github.com/slemeur/service-binding-documentation/raw/main/static/resources/init-database.sh | bash
````
This requires an init container. I would prefer not to do this now as it is a complex setup.

baijum commented on docs/userguide/docs/getting-started/quick-start.md:
```diff
@@ -0,0 +1,342 @@
+
+## Connect the application to the database
+
+Suppose the Service Binding operator is not present. In that case, the
+application's admin needs to extract all the configuration details and create a
+Secret resource and expose it to the application through volume mount in
+Kubernetes. The steps would be something like this:
```
Updated text

baijum commented on docs/userguide/docs/getting-started/quick-start.md:
```diff
@@ -0,0 +1,342 @@
+that, we'll need to create a Service Binding ressource which will trigger the
+Service Binding Operator to project the binding metadatas into the application.
+
+1. Create the ServiceBinding custom resource to project the binding metadata:
```
Done

baijum commented on docs/userguide/docs/getting-started/quick-start.md:
```diff
@@ -0,0 +1,342 @@
+ServiceBinding resource, there are two service resources, one custom resource
+and another Secret resource. This is required as the values required for
+database connectivity are living in these two resources. The application
+points to a `Deployment` or any similar resource with an embedded `PodSpec`.
```
Done

baijum commented on docs/userguide/docs/getting-started/quick-start.md:
```diff
@@ -0,0 +1,342 @@
+By creating this `Service Binding` resource, we now have values from the
+database's binding metadata projected into the application container as files
+(that's the default behavior, but you can also project environment variables if
+you prefer). If you check under `/bindings/spring-petclinic-rest` directory
+you'll see all the values from the secret resource projected there. In the above
+example, you'll find `username` and `password`. And the values pointed out
```
This check will help the user to understand the projection better.

baijum commented on docs/userguide/docs/getting-started/quick-start.md:
```diff
@@ -0,0 +1,342 @@
+
+In this quick start, we are going to leverage Service Binding as a way to easily
+and safely connect the application to the database service. In order to do
+that, we'll need to create a Service Binding ressource which will trigger the
```
Fixed

baijum commented on docs/userguide/docs/getting-started/quick-start.md:
```diff
@@ -0,0 +1,342 @@
+/bindings/spring-petclinic-rest/username: hippo
+/bindings/spring-petclinic-rest/password: w0ZB<0j1W|K;+4*TlK7-w^z/
+/bindings/spring-petclinic-rest/host: hippo-pgbouncer
+/bindings/spring-petclinic-rest/port: 5432
```
This check will help the user to understand the projection better.

baijum commented on docs/userguide/docs/creating-service-bindings/creating-service-binding.md:
```diff
@@ -0,0 +1,80 @@
+5. Detect Bindings through owned resources
+
+You also required the application ready before creating the ServiceBinding
+resource.
```
Rephrased the sentence. The "application" should exist.

baijum commented on docs/userguide/docs/creating-service-bindings/creating-service-binding.md:
```diff
@@ -0,0 +1,80 @@
+There are two APIs supported by the Service Binding Operator.  Both has
+`ServiceBinding` as the same name for the kind.  But the API group is different.
+
+1. `binding.operators.coreos.com`
```
Done

baijum commented on docs/userguide/docs/creating-service-bindings/creating-service-binding.md:
```diff
@@ -0,0 +1,80 @@
+`ServiceBinding` as the same name for the kind.  But the API group is different.
+
+1. `binding.operators.coreos.com`
+2. `service.binding` (API group used in the [Service Binding Spec][spec])
```
Done.

baijum commented on docs/userguide/docs/creating-service-bindings/creating-service-binding.md:
```diff
@@ -0,0 +1,80 @@
+| Support to auto-detect binding resources | Does not support auto-detecting binding resources |
+| Flag to bind as files or env vars        | No flag switch between files and env vars         |
+
+Here is the example configuration with `binding.operators.coreos.com` API group:
```
Done

baijum commented on docs/userguide/docs/creating-service-bindings/creating-service-binding.md:
```diff
@@ -0,0 +1,80 @@
+  name: account-service
+spec:
+  application:
+    apiVersion: apps/v1
```
Fixed.

baijum commented on docs/userguide/docs/creating-service-bindings/creating-service-binding.md:
```diff
@@ -0,0 +1,80 @@
+    name: online-banking
+
+  services:
+  - apiVersion: example.com/v1alpha1
```
Fixed.

baijum commented on docs/userguide/docs/creating-service-bindings/binding-options.md:
```diff
@@ -0,0 +1,174 @@
+
+## Transform binding names before projection into application
+
+WARNING: This feature is only available for `ServiceBinding` in the
```
Changed to note.

baijum reviewed (commented):
> A lot of the commands reference https://github.com/slemeur/service-binding-documentation this should be updated to reflect the new location.

Updated the links

baijum commented:
> spell check should be run on the doct

Done.

> replace in general "binding metadata" with "binding data"

Done.

> replace "service binding custom resource" with "service binding resource"

Done.

> spec API meanwhile got updated to servicebinding.io and version is v1alpha3

Done.

baijum commented on docs/userguide/docs/intro.md:
```diff
@@ -0,0 +1,71 @@
+[spec]: https://github.com/k8s-service-bindings/spec
+
+Note: There are two CRDs with name `ServiceBinding`.  The API group
+`service.binding` is used for the spec based CRD whereas
```
Update to new API group.

baijum commented on docs/userguide/docs/creating-service-bindings/binding-options.md:
````diff
@@ -0,0 +1,174 @@
+projected environment variable is going to be like this:
+
+```yaml
+DATABASE_HOST: example.com
````
Updated. See line 32.

baijum commented on docs/userguide/docs/creating-service-bindings/binding-options.md:
````diff
@@ -0,0 +1,174 @@
+POSTGRES_DATABASE_PORT_ENV: 8080
+```
+
+We can define how that key should be prepared defining string template in `namingStrategy`
````
Changed `that` to `each`.

baijum commented on docs/userguide/docs/creating-service-bindings/binding-options.md:
```diff
@@ -0,0 +1,174 @@
+
+We can define how that key should be prepared defining string template in `namingStrategy`
+
+#### Naming Strategies
```
Moved this before ` Compose custom binding data` section.

baijum commented on docs/userguide/docs/creating-service-bindings/binding-options.md:
```diff
@@ -0,0 +1,174 @@
+
+There are few naming strategies predefine.
+
+1. `none` - When this is applied, we get binding names in following form - `{{ .name }}`
```
Fixed.

baijum commented on docs/userguide/docs/creating-service-bindings/service-binding-resource.md:
```diff
@@ -0,0 +1,84 @@
+sidebar_position: 6
+---
+
+# Service Binding Resource
```
A link is already given (line 9). Please confirm, if we can remove the text.

pedjak replied:
I would remove it.

baijum commented on docs/userguide/docs/exposing-binding-data/intro-expose-binding.md:
```diff
@@ -0,0 +1,86 @@
+    `Secret` that you can use to detect the binding metadata.
+
+Service Binding Operator provides the ability to expose the values from the
+backing service resources and CRDs.  You can use the following methods to map
```
Done.

baijum commented on docs/userguide/docs/using-projected-bindings/using-projected-bindings.md:
```diff
@@ -0,0 +1,124 @@
+
+The primary mechanism of projection is through files mounted at a specific
+directory.  The binding metadata directory path can be discovered through
+`SERVICE_BINDING_ROOT` environment variable setup in your application.
```
Done

baijum commented on docs/userguide/docs/using-projected-bindings/using-projected-bindings.md:
```diff
@@ -0,0 +1,124 @@
+
+### Environment Variables
+
+The spec also has support for projecting environment variables.  You can use the
```
Done.

slemeur commented on docs/userguide/docs/creating-service-bindings/creating-service-binding.md:
```diff
@@ -0,0 +1,80 @@
+You also required the application ready before creating the ServiceBinding
+resource.
+
+There are two APIs supported by the Service Binding Operator.  Both has
+`ServiceBinding` as the same name for the kind.  But the API group is different.
```
Service Binding Operator is providing and supporting two different group APIs. 
-  `binding.operators.coreos.com/v1alpha1`: This group API is compliant with the Service Binding specification. It is fully supported and considered as the mature API. It's the group API we recommend in production environments.
- `servicebinding.io/v1alpha3` : This group API is the one defined in the Service Binding specification. The specification is still evolving and maturing, as a result the group API might change in the future. We recommend you to carefully use that group API for testing purposes and preparing your integration to become fully compliant with the specification. It is available in Service Binding Operator 1.x, as Tech Preview. 

We'll consider supporting only the group API from the specification, once the Service Binding specification will be officially GA. Of course, that will be a breaking change and handle in a safe way for you.

WDYT @pedjak ?

pedjak replied:
sounds good - I would just replace "group API" with API for clearness.

slemeur replied:
sounds good ! Thanks @pedjak

baijum replied:
Updated text.

baijum commented on docs/userguide/docs/intro.md:
```diff
@@ -0,0 +1,71 @@
+
+[spec]: https://github.com/k8s-service-bindings/spec
+
+Note: There are two CRDs with name `ServiceBinding`.  The API group
```
removed

sadlerap reviewed (commented):
Some minor spelling/syntax/grammar rewordings and cleanups.  This should be mostly low-hanging fruit.

If you think these changes aren't relevant for getting this merged, let me know, and I'll file a separate PR when this gets merged.
